### ISSUE TITLE
[registration] A problem in function computeTransformation of pcl::PPFRegistration

### ISSUE THREAD

**Issue by kaiwen2020** [PointCloudLibrary/pcl#4936]
There might be a problem in the computeTransformation function of pcl::PPFRegistration at line 157

```c++
unsigned int alpha_discretized = static_cast<unsigned int>(
                std::floor(alpha) +
                std::floor(M_PI / search_method_->getAngleDiscretizationStep()));
```

Since in PPFRegistration, the aim of this step is to discretize the angle alpha, should not it be
```c++
unsigned int alpha_discretized = static_cast<unsigned int>(
                std::floor((alpha + M_PI) /  search_method_->getAngleDiscretizationStep()) ;
```
 - PCL Version [1.12.0]

I'm not sure if I understand it right. Can anyone check this line since it might affect the matching performance? Thank you.

**Comment by mvieth** [comment]
I am not very familiar with this class, but I agree that those lines look wrong. Could you explain your suggested fix a bit more?
I had a quick look at [the original paper](http://far.in.tum.de/pub/drost2010CVPR/drost2010CVPR.pdf), but didn't find anything on how they discretize the alpha.
While investigating, I also saw that `PPFRegistration` is currently not tested: [there is a test](https://github.com/PointCloudLibrary/pcl/blob/master/test/registration/test_registration.cpp#L727), but that is disabled due to some problem. So that could be a hint that this line is wrong and messes up the calculation.

**Comment by mvieth** [comment]
It seems like this has been discovered before: #1930

**Comment by kaiwen2020** [comment]
Thanks for your reply and for additional information, I checked  this issue #1930, since the range that atan returns is between -pi/2 and pi/2 radians, the range of alpha is between -pi and pi. I think what I suggest is right
```c++
unsigned int alpha_discretized = static_cast<unsigned int>(
                std::floor((alpha + M_PI) /  search_method_->getAngleDiscretizationStep()) ;
```

I didn't find  in the article about the discretization step of alpha either, only this line
> The number of columns Nangle corresponds to the number of sample steps nangle of the rotation angle α.

**Comment by mvieth** [comment]
> I checked this issue #1930, since the range that atan returns is between -pi/2 and pi/2 radians, the range of alpha is between -pi and pi. In this way what I suggest above should be right.

`alpha_s` is computed with `atan2`:
https://github.com/PointCloudLibrary/pcl/blob/ce8d85d1237bd0e8ac2d94446a43159adf00c818/registration/include/pcl/registration/impl/ppf_registration.hpp#L143-L144
`atan2` returns values between -pi and +pi ([see here](https://en.cppreference.com/w/cpp/numeric/math/atan2)). alpha_m should be in the same range, so `alpha` (computed by `alpha = alpha_m - alpha_s`) would be in the range from `-2*pi` to `+2*pi`

Another thing: we should also check whether these lines are correct, they seem suspicious too:
https://github.com/PointCloudLibrary/pcl/blob/ce8d85d1237bd0e8ac2d94446a43159adf00c818/registration/include/pcl/registration/impl/ppf_registration.hpp#L203-L207

**Comment by kaiwen2020** [comment]
You are right, Lanns's answer is correct, we should constrain alpha to the range -pi to pi. The  suspicious lines seem correct, since during discretization, we add M_PI to alpha to assure the index of a vector to be positive or 0, then when we take the value of alpha from the accumulator array, we should substract the index with 
`std::floor(static_cast<float>(M_PI) / 
                               search_method_->getAngleDiscretizationStep()`

**Comment by mvieth** [comment]
I am still not sure about three things:
1. Why does the `floor` make sense? That would change e.g. 19.99999 to 19.0. And since the angle should be a float anyway, I don't see a reason to use `floor`. Do you have an explanation?
2. The fact that the discretization step is first divided by, and later multiplied by. If the `floor` wasn't there, that could be simplified
3. I think it should be `max_votes_j+0.5` instead of just `max_votes_j`. E.g. for `max_votes_j=0`, we would be in the bin from `-pi` to `-pi+search_method_->getAngleDiscretizationStep()`. But the bin is better represented by the centre of the bin (`-pi+0.5*search_method_->getAngleDiscretizationStep()`), than by the left side (`-pi`)

**Comment by mvieth** [comment]
@kaiwen2020 Do you have any further ideas/thoughts about this? What do you think about the three things I described above?

**Comment by kaiwen2020** [comment]
max_votes_j is actually an int, since it is obtained this way `max_votes_j = j`. It is actualy an index of the accumulator array. And this `std::floor(static_cast<float>(M_PI) /search_method_->getAngleDiscretizationStep())` is to discretize the M_PI in the same way as alpha. We can consider it as a shift in the accumulator array, it should be an integer since it is an index. I think it is better to use

**Comment by kaiwen2020** [comment]
You are right. Theoretically `floor` works since the discretization step is pi divided by an integer; but numericallly, `floor` may be wrong due to the precision of floating point. I did not consider this... Maybe it is more reasonable to use `round` here? Also I agree that it is better to represent an interval by its center instead of the left side.

**Comment by mvieth** [comment]
@kaiwen2020 Would you be interested in opening a pull request to fix the problems we discussed?

**Comment by kaiwen2020** [comment]
I'm willing to, I will read the tutorial first [https://github.com/PointCloudLibrary/pcl/wiki/A-step-by-step-guide-on-preparing-and-submitting-a-pull-request](url) as suggest in issue #1930 since I haven't contributed to PCL before

**Comment by mvieth** [comment]
@kaiwen2020 If you have any questions about the process, feel free to ask

**Comment by kaiwen2020** [comment]
@mvieth I have a question when modifying the code, from line 143 to 147 in ppf_registration.hpp
```c++
 float alpha_s =
              std::atan2(-scene_point_transformed(2), 
  if (std::sin(alpha_s) * scene_point_transformed(2) < 0.0f)
            alpha_s *= (-1);
  alpha_s *= (-1);
```
and also from line 96 to 99 in ppf.hpp
```c++
float angle = std::atan2 ( -model_point_transformed(2), model_point_transformed(1));
if (std::sin (angle) * model_point_transformed(2) < 0.0f)
       angle *= (-1);
 p.alpha_m = -angle;
```
What's the intention of these lines... to force alpha_s and alpha_m to be positive? If so, then the range of alpha should be from -pi to pi. What's your opinion?
